### PR TITLE
Server side purge

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Adds serverside pseudo purge which runs on a schedule and executes a configurable "purge" function over all docs that would be replicated. "Purged" docs are added in auxiliary databases - one for every unique user roles group.
During offline user replication (both initial and subsequent), only docs that are not tagged as "purged" are downloaded. On a schedule, on application bootstrap, the webapp will check if new documents were "purged" and delete them from the device.

medic/medic#5443

# Long description

Executes a pseudo server-side purge which:
- runs on a configurable schedule
- computes a list of unique role groups from all existent users, each unique role group purges being saved in a separate database
- compiles groups consisting of: a contact, reports _about_ the contact, messages that the contact has sent or received
- over each doc group and for each role group, runs a configurable purge function which accepts the group + the roles as parameters and is expected to return an array of doc_ids which should be purged
- queries the purge dbs and compare the resulting list of purged ids compared to the existent purged docs and saves the differences (deleting old purges or by adding new)
- a doc is purged for a list of roles if a document with the same id, prefixed with `purge:`, exists in the corresponding purge database (eg. `purge:this_is_my_doc_id`)

When performing an initial replication:
- users will only download changes for docs that are not purged for their role set 
- the list of purged doc ids is cached for 5 minutes (with refreshed ttl)
- after initial replication, webapp calls an api endpoint which will create a "checkpointer" doc in the corresponding purge db that saves the current `update_seq`

When performing any replication:
- users will only download changes od docs that are not purged for their role set

On bootstrap, every 7 days or configurable via `purge: run_every_days` or when user roles change: 
- webapp calls an api endpoint which will query _changes from the corresponding purge db from the seq saved in the checkpointer doc
- the endpoint returns a list of new purged doc ids and the a seq of the purge db
- webapp deletes the docs locally, also marking them as purged
- webapp queries an api endpoint that updates the purge checkpointer doc using the seq received in the previous response

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
Reasoning behind:
1. purging per role set:
  - it's closer to the way purge is described now
  - compared to running purge per one role (with separate dbs), it's faster when replicating because only one db is queried
  - compared to running purge per one role (with singular db), it's faster when replicating because we don't need to multiply the allowed doc ids with the number of roles the user has when getting purges

2. separate purge dbs per role set:
  - possibility for fast cleanup: when a role group is no longer used, we can just delete the whole purge database
  - when a user's roles change, the purge updating is done naturally via the missing checkpointer doc in the corresponding db

**Comment by dianabarsan** [comment]
Hey! Thanks a lot for the overview.

> 1. Instead of providing purge APIs and handling checkpointing and so on, would it be better/easier to do a filtered replication of the purge db?

I implemented this when the purges where done per single role and users would get the purged docs from a list of databases corresponding to their roles. Using PouchDB replication wasn't an option in that case. 
We would have to write a custom _changes controller for this replication, filtering client side by doc_ids will generate a different checkpoint hash every time the local doc_ids change. We would have to open the purge databases to _local writes from all users, along with the other endpoints needed to for replication. 
Clients would replicate full docs, which should be immediately thrown away instead of saved and waste storage space on the device, plus the replication would be a little chattier (with _changes and _all_docs requests). 
But I agree that using PouchDB replication would be possible with the current implementation.
Do you think it's worth the downsides? 

> 2. I don't think you've removed the client side purging - were you intending to keep that around as well?

That is correct, I have not removed it *yet*. Ideally I would, but wanted to discuss it before I did. This is confirmation that I should? :grin: 

> 3. I wonder if checking for purged docs after each successful replication is too chatty - can we do it less often, say once a month at startup (just like the client side purging code)?

That would be fine by me. I could straight up replace the local purge code to replicate from the purge db instead of running local purge. 

> 4. You're using arrays of roles everywhere but user's can only have one role of list defined in `roles` object of the settings doc. The other roles are irrelevant. If you find the one we care about and just use that I think it would simplify the code somewhat.

That is, unfortunately, not the case. I have been looking at users from various projects and found near to none that use a single role from the configured roles, most use two or more. For example this muso user: 

```
{
  "_id": "org.couchdb.user:anonymous",
  "type": "user",
  "name": "anonymous",
  "roles": [
    "district-manager",
    "kujua_user",
    "data_entry",
    "supervisor",
    "mrdt_tmp",
    "supervisor_mrdt",
    "branch_manager"
  ],
  "facility_id": "...",
  "known": true
}
```

Where the `settings` doc has the following roles configured: national_admin, district_admin, **data_entry**, analytics, gateway, chw_uhc, chw_test_analytics, **supervisor**, **supervisor_mrdt**, **mrdt_tmp**, mrdt_tmp_2, pm, muso-sih - I bolded the roles that the user has and are configured. Many of these roles are also set as offline so there's no actual way of selecting a single *main* role. 

> 5. Why are you batching the purging by the root contact?

It was a way of providing a "`key`" param in the `contacts_by_depth` queries. In order to use `startkey_docid`. Although I could use `medic-client/contacts_by_type` instead. I wanted to use the same views as replication, but `contact_by_depth` emits multiple times per contact. Would querying `medic-client/contacts_by_type`  better?

**Comment by garethbowen** [comment]
> > Instead of providing purge APIs and handling checkpointing and so on, would it be better/easier to do a filtered replication of the purge db?

> Clients would replicate full docs, which should be immediately thrown away instead of saved and waste storage space on the device, plus the replication would be a little chattier (with _changes and _all_docs requests).

Yeah that's probably a big deal. Leave it as it is.

> > I don't think you've removed the client side purging - were you intending to keep that around as well?

> That is correct, I have not removed it yet. Ideally I would, but wanted to discuss it before I did. This is confirmation that I should?

I strongly think we should remove it. Because purging is run so infrequently it's fine to have a small delay while the user is offline and then do the purging when they next come online.

> > I wonder if checking for purged docs after each successful replication is too chatty - can we do it less often, say once a month at startup (just like the client side purging code)?

> That would be fine by me. I could straight up replace the local purge code to replicate from the purge db instead of running local purge.

Nice!

> > You're using arrays of roles everywhere but users can only have one role of list defined in roles object of the settings doc. The other roles are irrelevant. If you find the one we care about and just use that I think it would simplify the code somewhat.

> That is, unfortunately, not the case.

Oh wow. Can you please raise an issue to discuss whether that should be supported or not? If it's supported we should make sure the codebase works with it, and update the UI to make it possible to configure.

> > Why are you batching the purging by the root contact?

> It was a way of providing a "key" param in the contacts_by_depth queries. In order to use startkey_docid. Although I could use medic-client/contacts_by_type instead. I wanted to use the same views as replication, but contact_by_depth emits multiple times per contact. Would querying medic-client/contacts_by_type better?

Yes i think it would be cleaner to use `contacts_by_type`.

**Comment by dianabarsan** [comment]
> > > I don't think you've removed the client side purging - were you intending to keep that around as well?
> 
> > That is correct, I have not removed it yet. Ideally I would, but wanted to discuss it before I did. This is confirmation that I should?
> 
> I strongly think we should remove it. Because purging is run so infrequently it's fine to have a small delay while the user is offline and then do the purging when they next come online.

> > > I wonder if checking for purged docs after each successful replication is too chatty - can we do it less often, say once a month at startup (just like the client side purging code)?
> 
> > That would be fine by me. I could straight up replace the local purge code to replicate from the purge db instead of running local purge.
> 
> Nice!
> 

Cool, I'm in the process of making these 2 changes above.

> > > You're using arrays of roles everywhere but users can only have one role of list defined in roles object of the settings doc. The other roles are irrelevant. If you find the one we care about and just use that I think it would simplify the code somewhat.
> 
> > That is, unfortunately, not the case.
> 
> Oh wow. Can you please raise an issue to discuss whether that should be supported or not? If it's supported we should make sure the codebase works with it, and update the UI to make it possible to configure.
> 

Will do. 

> > > Why are you batching the purging by the root contact?
> 
> > It was a way of providing a "key" param in the contacts_by_depth queries. In order to use startkey_docid. Although I could use medic-client/contacts_by_type instead. I wanted to use the same views as replication, but contact_by_depth emits multiple times per contact. Would querying medic-client/contacts_by_type better?
> 
> Yes i think it would be cleaner to use `contacts_by_type`.

Great, already changed.

**Comment by dianabarsan** [comment]
> I wonder if checking for purged docs after each successful replication is too chatty - can we do it less often, say once a month at startup (just like the client side purging code)?

I just re-realized why I chose to only purge after a successful downwards replication. 
If the local db is not fully replicated down when I get the purged docs, we might risk documents that are not downloaded yet to become "purged" before they actually exist. That would make the purged rev likely be a losing rev and the doc to appear to the user permanently. 
Do you think this is an acceptable trade-off to purge on startup? 

I could keep a list purged-ids locally and filter out those changes on downwards replication, but I don't like that very much.

File: sentinel/src/config.js [hunk 1]
```diff
@@ -42,7 +42,11 @@ const initFeed = () => {
     .on('change', change => {
       if (change.id === 'settings') {
         logger.info('Reloading configuration');
-        initConfig();
+        initConfig().then(() => {
+          if (config['reschedule']) {
+            require('./schedule').checkSchedule();
```
Comment (dianabarsan) [inline sentinel/src/config.js]: Had to quickly devise a way of restarting the scheduler on the fly for e2e tests. Stefan has a PR that adds support to stop/restart api or sentinel in e2e, so once that is merged I plan to remove this and restart sentinel when I need it to re-schedule.

File: sentinel/src/schedule/index.js [hunk 2]
```diff
@@ -34,8 +35,11 @@ exports.sendable = function(config, now) {
   return now >= after && now <= until;
 };
 
+let timeout;
+
 exports.checkSchedule = function() {
   const now = moment(date.getDate());
+  clearTimeout(timeout);
```
Comment (dianabarsan) [inline sentinel/src/schedule/index.js]: This change is made to support the scheduler restart mentioned earlier. Will be removed once it's possible to restart sentinel in e2e tests.

**Comment by garethbowen** [comment]
> If the local db is not fully replicated down when I get the purged docs, we might risk documents that are not downloaded yet to become "purged" before they actually exist. That would make the purged rev likely be a losing rev and the doc to appear to the user permanently. Do you think this is an acceptable trade-off to purge on startup?

What about after successful replication we check to see if purge has run in the last month (or week?). If it has, don't run it again.

> I could keep a list purged-ids locally and filter out those changes on downwards replication, but I don't like that very much.

I don't like this much either...

> On second thought, I could filter out purged ids server-side all the time, but that makes the changes request take longer.

This is quite appealing because it saves you from downloading and purging but it's unlikely to happen often - if the doc has only just been created it probably won't also be purged. How much of an impact would this have on request performance?

**Comment by dianabarsan** [comment]
>> On second thought, I could filter out purged ids server-side all the time, but that makes the changes request take longer.

> This is quite appealing because it saves you from downloading and purging but it's unlikely to happen often - if the doc has only just been created it probably won't also be purged. How much of an impact would this have on request performance?

Obviously the impact is in direct proportion with the size of the database and the number of docs the user has, it's impossible to assign an exact value. 
I ran a small test on my test DB with:
- 5.5kk docs, 1kk persons, 300k clinics, 4kk reports
- user with 20k docs
- requesting with limit = 1

| Nbr docs | Nbr purged      | 1st request (sec)                    | AVG  | next requests (sec)            | AVG  |
|----------|-----------------|------------------------------------|------|------------------------------------|------|
| 20k      | 15k             | 3.29, 3.45, 3.10, 3.29, 3.30, 3.49 | 3.32 | 2.18, 2.26, 2.35, 2.22, 2.28, 2.05 | 2.21 |
| 20k      | 0               | 2.23, 2.32, 2.38, 2.35, 2.36, 2.30 | 2.32 | 2.20, 2.16, 2.21, 2.16, 2.22, 2.20 | 2.19 |
| 20k      | no purge filter |  -                                 | -    | 1.92, 2.27, 2.30, 2.20, 2.22, 2.21 | 2.18 |

Because the purged ids are cached, only the 1st request will be affected.

File: api/src/services/cache.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const NodeCache = require('node-cache');
+
+const opts = {
+  stdTTL: 5 * 60, // standard ttl in seconds for every generated cache element: 5 minutes
+  checkperiod: 5 * 60, // period in seconds used for the automatic delete check interval: 5 minutes
+  useClones: false
+};
+const cache = new NodeCache(opts);
+
+module.exports = cache;
```
Comment (garethbowen) [inline api/src/services/cache.js]: I'm not sure if this should be a service in its current form, because it always returns the same cache instance. If we decide to reuse it for something else it would probably be better to have a different cache so we don't accidentally collide on keys.

I guess you did it because it makes unit testing easier? If so, I'd change it so it exported a function which creates and returns new cache which then gives you a function to mock.

File: api/src/services/purged-docs.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+};
+
+let inited = false;
+const init = () => {
+  if (inited) {
+    return;
+  }
+  listen();
+  inited = true;
+};
```
Comment (garethbowen) [inline api/src/services/purged-docs.js]: Instead of an `init` function can we start the listener immediately when the module is `require`d? Again it makes testing a little harder because you have to mock before you require but it makes the code much cleaner because it'll only be invoked once and we never have to check again.

File: api/src/services/purged-docs.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+const getCacheKey = (roles, docIds) => {
+  const hash = crypto
+    .createHash('md5')
+    .update(JSON.stringify(docIds), 'utf8')
+    .digest('hex');
+
+  return `purged-${JSON.stringify(roles)}-${hash}`;
+};
+
+const regex = new RegExp(/^purged-.+-.+$/);
```
Comment (garethbowen) [inline api/src/services/purged-docs.js]: This could use a more descriptive name. Also, I think you're passing a regex to the Regex constructor which is unnecessary?

File: api/src/routing.js [hunk 6]
```diff
@@ -397,6 +398,12 @@ app.get('/api/v1/settings', settings.get);
 app.putJson(`${appPrefix}update_settings/${environment.ddoc}`, settings.put); // deprecated
 app.putJson('/api/v1/settings', settings.put);
 
+const purgedDocsController = require('./controllers/purged-docs');
```
Comment (garethbowen) [inline api/src/routing.js]: Can we put this at the top with the other `require` statements?

File: api/src/middleware/authorization.js [hunk 7]
```diff
@@ -23,6 +23,7 @@ module.exports = {
       .getUserCtx(req)
       .then(userCtx => {
         req.userCtx = userCtx;
+        req.replicationId = req.headers['medic-replication-id'];
```
Comment (garethbowen) [inline api/src/middleware/authorization.js]: This is only used in one place as far as I can see - instead of setting it in the middleware can we just access the header in the controller? That way it won't execute for non-purging requests.

File: ddocs/medic-client/views/contacts_by_type/map.js [hunk 8]
```diff
@@ -1,4 +1,9 @@
 function(doc) {
+  var tombstone = false;
+  if (doc.type === 'tombstone' && doc.tombstone) {
+    tombstone = true;
+    doc = doc.tombstone;
+  }
```
Comment (garethbowen) [inline ddocs/medic-client/views/contacts_by_type/map.js]: This seems like it may have unintended consequences, like, online users will start seeing tombstoned contacts. Why is this needed?

File: sentinel/src/lib/purging.js [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+            }
+            toPurge[hash][id] = true;
+          });
+        });
+      });
+
+      return updatePurgedDocs(rolesHashes, docIds, alreadyPurged, toPurge);
+    })
+    .then(() => nextKey && batchedContactsPurge(roles, purgeFn, nextKey, nextKeyDocId));
+};
```
Comment (garethbowen) [inline sentinel/src/lib/purging.js]: This function is too long - is there any way to break it up into separate functions with descriptive names?

File: shared-libs/purging-utils/src/index.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+const sortedUniqueRoles = roles => ([...new Set(roles)].sort());
+const getRoleHash = roles => crypto
+  .createHash('md5')
+  .update(JSON.stringify(sortedUniqueRoles(roles)), 'utf8')
+  .digest('hex');
+
+const getPurgeDbName = (mainDbName, rolesHash) => `${mainDbName}-purged-role-${rolesHash}`;
+
+// todo consider moving this somewhere else, it's a duplicate of a function that atm exists in the users controller in api
```
Comment (garethbowen) [inline shared-libs/purging-utils/src/index.js]: A new user shared-lib might work, but feel free to raise a separate issue for this.

File: webapp/src/js/bootstrapper/index.js [hunk 11]
```diff
@@ -164,17 +176,21 @@
...
+          .then(shouldPurge => {
+            if (!shouldPurge) {
+              return;
+            }
+
+            return serverSidePurge
+              .purge(localDb, userCtx)
+              .on('start', () => setUiStatus('PURGE_INIT'))
+              .on('progress', progress => setUiStatus('PURGE_INFO', { count: progress.purged }))
+              .catch(console.error);
```
Comment (garethbowen) [inline webapp/src/js/bootstrapper/index.js]: Add a string prefix to make it easier to find where the error was thrown after it's minified, eg:
`.catch(err => console.error('Error attempting to purge', err));`

File: webapp/src/js/bootstrapper/index.js [hunk 12]
```diff
@@ -32,46 +33,54 @@
...
+            timeout: 1000 * 60 * 10, // try for ten minutes then give up,
+            query_params: { initial_replication: true }
+          });
 
-    return replicator
-      .then(function() {
-        var duration = Date.now() - dbSyncStartTime;
+        replicator
+          .on('change', function(info) {
+            console.log('initialReplication()', 'change', info);
```
Comment (garethbowen) [inline webapp/src/js/bootstrapper/index.js]: Accidentally left behind?

File: webapp/src/js/bootstrapper/server-side-purge.js [hunk 13]
```diff
@@ -0,0 +1,170 @@
+const utils = require('./utils');
```
Comment (garethbowen) [inline webapp/src/js/bootstrapper/server-side-purge.js]: Maybe this file should just be called `purger.js` now that we only have one?

File: webapp/src/js/bootstrapper/utils.js [hunk 14]
```diff
@@ -13,8 +13,16 @@ const feedback = msg => {
   }
 };
 
+const getBaseUrl = () => {
+  // parse the URL to determine the remote and local database names
+  const location = window.location;
+  const port = location.port ? ':' + location.port : '';
+  return `${location.protocol}//${location.hostname}${port}`;
+};
```
Comment (garethbowen) [inline webapp/src/js/bootstrapper/utils.js]: Another shared lib waiting to happen...

File: ddocs/medic-client/views/contacts_by_type/map.js (hunk 8 above)
Comment (dianabarsan) [reply ddocs/medic-client/views/contacts_by_type/map.js]: This is needed so we purge reports about deleted contacts. This was another reason why I chose querying `contacts_by_depth` initially. 
I don't believe online users will start seeing deleted contacts, in webapp the view is always called with specific keys and tombstones emit a prefixed key ("tombstone-person"). The only place where the view is called without keys is in exports. 
Do you think it would be worth to create a new view in `medic` for this purpose?

File: webapp/src/js/bootstrapper/index.js (hunk 12 above)
Comment (dianabarsan) [reply webapp/src/js/bootstrapper/index.js]: Argh, you made me doubt myself! No, not left behind, part of `master` code. 
https://github.com/medic/medic/blame/4c70aca3a3024cd6e233ffb3b74c72d8bfb0b02b/webapp/src/js/bootstrapper/index.js#L64

File: webapp/src/js/bootstrapper/server-side-purge.js (hunk 13 above)
Comment (dianabarsan) [reply webapp/src/js/bootstrapper/server-side-purge.js]: Yea, will do. I had 2 files at some point (the old and the new).

File: api/src/middleware/authorization.js (hunk 7 above)
Comment (dianabarsan) [reply api/src/middleware/authorization.js]: I had that idea some time ago about making an initial replication cache. The first step was to consistently and uniquely identify the user + device they were replicating from.
This is the hole this header wanted to fill. 

If you don't like it, since purging is calling it's own endpoints, I can easily just abandon the header and use a query parameter.

File: api/src/services/cache.js (hunk 3 above)
Comment (dianabarsan) [reply api/src/services/cache.js]: Switched it to return a named cache instance.

File: api/src/services/purged-docs.js (hunk 4 above)
Comment (dianabarsan) [reply api/src/services/purged-docs.js]: Done. I just added an `if` not to start the listener when running tests.

File: api/src/services/purged-docs.js (hunk 5 above)
Comment (dianabarsan) [reply api/src/services/purged-docs.js]: Removed as no longer needed. With the separate cache instances, we clear all keys when purge runs.

File: api/src/routing.js (hunk 6 above)
Comment (dianabarsan) [reply api/src/routing.js]: Done

File: sentinel/src/lib/purging.js (hunk 9 above)
Comment (dianabarsan) [reply sentinel/src/lib/purging.js]: I refactored it but I don't think I did a good job naming those auxiliary functions. Suggestions welcome!

File: shared-libs/purging-utils/src/index.js (hunk 10 above)
Comment (dianabarsan) [reply shared-libs/purging-utils/src/index.js]: Will do after that other PR is merged.

File: webapp/src/js/bootstrapper/index.js (hunk 11 above)
Comment (dianabarsan) [reply webapp/src/js/bootstrapper/index.js]: Done.

**Comment by dianabarsan** [comment]
Additionally to your requested changes, I also added one update to the "shouldPurge" function client-side. 
There was a piece of code that I took from the previous `purger`, where it would postpone running purge if the db was not synced. 
I removed it because I wanted purge to be resumable when refreshing the page: since purge works by deleting documents locally, even purging one doc would cause the DB to be out of sync, so purge would not run on next load. It would have required a second refresh to resume purging. 
I'm also going to add a PR to update `medic-docs`.

File: webapp/src/js/bootstrapper/purger.js [hunk 15]
```diff
@@ -1,303 +1,108 @@
...
-        console.log('Local DB may have unreplicated records, skipping');
+const daysToMs = (days) => 1000 * 60 * 60 * 24 * days;
+module.exports.shouldPurge = (localDb, userCtx) => {
+  return Promise
+    .all([ localDb.get('settings'), getPurgeLog(localDb) ])
+    .then(([ { settings: { purge } }, purgelog ]) => {
+      // purge not running on the server
+      if (!purge) {
+        console.debug('Not purging: Purge not configured.');
+        return false;
```
Comment (dianabarsan) [inline webapp/src/js/bootstrapper/purger.js]: todo: don't start purge if device is not online!

**Comment by dianabarsan** [comment]
@garethbowen , as I was writing the documentation, I realized I forgot to do a connectivity check when running purge locally, so I added that, along with a check when the API endpoints are queried so they don't automatically create purge DB-s that don't exist yet. If you'd like to take a look.
Thanks a lot for your time!

**Comment by garethbowen** [comment]
Looks good. Nice one.

**Comment by dianabarsan** [comment]
Hey @garethbowen . Over the last couple of days I've added some updates, clearing some todos: 

- I rebased to master to get the `users-info` doc count endpoint and only return unpurged doc counts.
- Removed the hotfix to restart Sentinel scheduler in  Travis and straight-up copied Stefan's code from https://github.com/medic/medic/pull/5822
- Added the fix for old versions of fetch not defaulting to `same-origin` credentials.

**Comment by garethbowen** [comment]
Because it got force pushed it's difficult to review just what changed, but it all sounds good. Is there anything specifically you'd like me to review?

**Comment by dianabarsan** [comment]
These commits sum the changes up: https://github.com/medic/medic/pull/5878/commits/49779914df26c2b731f3880a5e3f7f9ab4ecd6de - ( though it's tainted by some post-rebase conflict resolutions) and https://github.com/medic/medic/pull/5878/commits/40d7b71604d9636da9524f0f77b396e9eecf54ec

I added a new exportable in `api/services/purged-docs` to return unpurged ids only - since now this was required in 2 places: the `change` controller and the `users` controller.

**Comment by garethbowen** [comment]
LGTM!

**Comment by dianabarsan** [comment]
@garethbowen 
Oh, there was something else I nearly forgot about.
@newtewt asked if there is a UI option to force the client to reload the app and download purge. 
What do you think? I was leaning towards some button in the "About" page?

**Comment by garethbowen** [comment]
I understand why QA would want it but I don't think it's useful for users, is it?

Maybe we need a new hidden page for engineers to do stuff like this. Feel free to leave it for another time.

**Comment by dianabarsan** [comment]
Thanks! 
To quote Nick directly: 
> Sometimes I wish the app had a developer console to force things.

**Comment by newtewt** [comment]
@dianabarsan  I'm going to leave this one to you to merge but I'm done with QA and no issues I can find.
